### PR TITLE
[core] Cleanup handling for nondeterministic object size during transfer

### PR DESCRIPTION
**Pull request by stephanie-wang**
## Why are these changes needed?

Currently object transfers assume that the object size is fixed. This is a bad assumption during failures, especially with lineage reconstruction enabled and tasks with nondeterministic outputs.

This PR cleans up the handling and hopefully guards against two cases where the object size may change during a transfer:
1. The object manager's size information does not match the object in the local plasma store (due to async notifications). --> the object manager overwrites its own information if it finds that the physical object has a different size.
2. The receiver's created buffer size does not match the sender's object size. --> the receiver destroys the previous buffer and creates a new buffer with the correct size. This might cause some transient errors but eventually object transfer should succeed.

Unfortunately I couldn't trigger this from Python because it depends on some pretty specific timing conditions. However, I did add some unit tests for case 2 (this is the majority of the PR).

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/object_manager/object_buffer_pool.cc [hunk 1]
```diff
@@ -20,11 +20,9 @@
 
 namespace ray {
 
-ObjectBufferPool::ObjectBufferPool(const std::string &store_socket_name,
-                                   uint64_t chunk_size)
-    : store_socket_name_(store_socket_name), default_chunk_size_(chunk_size) {
-  RAY_CHECK_OK(store_client_.Connect(store_socket_name_.c_str(), "", 0, 300));
-}
+ObjectBufferPool::ObjectBufferPool(
+    std::shared_ptr<plasma::PlasmaClientInterface> store_client, uint64_t chunk_size)
```
Comment (ericl) [inline src/ray/object_manager/object_buffer_pool.cc:24]: :+1:

File: src/ray/object_manager/object_buffer_pool.cc [hunk 2]
```diff
@@ -198,13 +212,31 @@ ray::Status ObjectBufferPool::EnsureBufferExists(const ObjectID &object_id,
...
+                          "nondeterministic output. Recreating object with size "
+                       << data_size << ".";
+      AbortCreateInternal(it->first);
+    }
+  }
+
   const int64_t object_size =
       static_cast<int64_t>(data_size) - static_cast<int64_t>(metadata_size);
   std::shared_ptr<Buffer> data;
+  RAY_LOG(INFO) << "store_client_ " << store_client_;
```
Comment (ericl) [inline src/ray/object_manager/object_buffer_pool.cc:235]: Stray log?

**Review by ericl — APPROVED** [review]
Nice! While this doesn't block this PR, I'm wondering if we may also see cases where the data size is the same but data contents differ across versions of the object.

I believe that can't happen since we currently stream object data from a single source (never re-using chunks), but we may want to add a random version / checksum of the object data to reject these cases as well in the future.

File: src/ray/object_manager/object_buffer_pool.cc (hunk 2 above)
Comment (stephanie-wang) [reply src/ray/object_manager/object_buffer_pool.cc:235]: Oops, thanks...

**Comment by stephanie-wang** [comment]
> Nice! While this doesn't block this PR, I'm wondering if we may also see cases where the data size is the same but data contents differ across versions of the object.
> 
> I believe that can't happen since we currently stream object data from a single source (never re-using chunks), but we may want to add a random version / checksum of the object data to reject these cases as well in the future.

Yeah I was thinking this as well, a version number would be good and I think it will work pretty much use the same codepath.

By the way, it is actually possible to get chunks from different sources right now if a transfer fails midway through, or if pull retries are close enough together that they overlap. That also means it's possible to get liveness issues if this happens repeatedly, but I figured it's fine for now.

**Comment by ericl** [comment]
> By the way, it is actually possible to get chunks from different sources right now if a transfer fails midway through, or if pull retries are close enough together that they overlap. That also means it's possible to get liveness issues if this happens repeatedly, but I figured it's fine for now.

Hmm, IIUC in this case the new source would at least push data for the chunks (overwriting previous content). Or do we ignore redundant chunks?

**Comment by ericl** [comment]
> By the way, it is actually possible to get chunks from different sources right now if a transfer fails midway through, or if pull retries are close enough together that they overlap. That also means it's possible to get liveness issues if this happens repeatedly, but I figured it's fine for now.

Ah, this is if there are two concurrent pushers to the same pull requester? That does sound problematic.

**Comment by scv119** [comment]
Yup if we add a unique version number (like a randomized UUID) should help up eliminate the same size different content case.

**Review by scv119 — APPROVED** [review]

**Pull request merged**